app/domain/authentication/authn_iam/authenticator.rb

```diff
@@ -0,0 +1,84 @@
+require 'active_support'
+require 'active_support/core_ext'
+require 'net/ldap'
+require 'json'
+require 'aws-sdk-iam'
+require 'aws-sdk-core'
+
+
+module Authentication
+  module AuthnIam
+
+    class Authenticator
+      
+      
+      def initialize(env:)
+        @env = env
+      end
+
+      def valid?(input)
+        @authn_name, @account, @login, password, @service_id = input.authenticator_name, input.account, input.username, input.password, input.service_id
+
+        # JSON holding the AWS signed headers 
+        @signed_aws_headers = JSON.parse input.password
+
+        is_trusted_by_aws? && (iam_role_matches? host_role)
+
+      end
+
+      def is_trusted_by_aws?
+
+        url = 'https://sts.amazonaws.com/?Action=GetCallerIdentity&Version=2011-06-15'
+
+        # Executes the AWS Signed Request
+        uri = URI.parse(url)
+        https = Net::HTTP.new(uri.host,uri.port)
+        https.use_ssl = true
+        https.set_debug_output($stdout)
+        request = Net::HTTP::Get.new(url)
```

Comment:
There's a lot of boilerplate network code here.  What do you think about pulling this out into a method.  A separate method would allow you mock the request option to validate the correct fields were added to the request object.  For example, we had something like:
```ruby
def base_aws_request
  url = 'https://sts.amazonaws.com/?Action=GetCallerIdentity&Version=2011-06-15'

  # Executes the AWS Signed Request
  uri = URI.parse(url)
  https = Net::HTTP.new(uri.host,uri.port)
  https.use_ssl = true
  https.set_debug_output($stdout)
  Net::HTTP::Get.new(url)
end

def retrieve_iam_identity(aws_headers)
  aws_request = base_aws_request
  aws_headers.each do |key, value|
    aws_request.add_field(key, value)
  end

  https.request(aws_request)
end
```
This would allow us to stub the `base_aws_request` to verify the expected header attributes were added.  It would also let you mock the AWS call to test this code made the call to AWS.

Reply:
Incorporated!